source/extensions/filters/http/dynamic_forward_proxy/proxy_filter.cc

```diff
@@ -0,0 +1,60 @@
+#include "extensions/filters/http/dynamic_forward_proxy/proxy_filter.h"
+
+#include "extensions/common/dynamic_forward_proxy/dns_cache.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace DynamicForwardProxy {
+
+ProxyFilterConfig::ProxyFilterConfig(
+    const envoy::config::filter::http::dynamic_forward_proxy::v2alpha::FilterConfig& proto_config,
+    Extensions::Common::DynamicForwardProxy::DnsCacheManagerFactory& cache_manager_factory,
+    Upstream::ClusterManager& cluster_manager)
+    : dns_cache_manager_(cache_manager_factory.get()),
+      dns_cache_(dns_cache_manager_->getCache(proto_config.dns_cache_config())),
+      cluster_manager_(cluster_manager) {}
+
+void ProxyFilter::onDestroy() {
+  // Make sure we destroy any active cache load handle in case we are getting reset and deferred
+  // deleted.
+  cache_load_handle_.reset();
+}
+
+Http::FilterHeadersStatus ProxyFilter::decodeHeaders(Http::HeaderMap& headers, bool) {
+  Router::RouteConstSharedPtr route = decoder_callbacks_->route();
+  if (!route || !route->routeEntry()) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  Upstream::ThreadLocalCluster* cluster =
+      config_->clusterManager().get(route->routeEntry()->clusterName());
+  if (!cluster) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  uint16_t default_port = 80;
+  if (cluster->info()->transportSocketFactory().implementsSecureTransport()) {
+    default_port = 443;
+  }
+
+  cache_load_handle_ =
+      config_->cache().loadDnsCache(headers.Host()->value().getStringView(), default_port, *this);
```

Comment:
This does work with the current implementation, and is tested in the integration tests. I will add more comments and docs on how the implementation handles ports.

Reply:
hm, I couldn't find it. the integration_test does only `localhost` test, no? I searched `loadDnsCache` calls and didn't find any IP address there. Also IPv6 address will break the split for host and port in `startCacheLoad`

Reply:
Oh sorry, nevermind I misread this. I will add a TODO to figure how to deal with this. Allowing IP:port seems bad since it will never work with TLS verification.

Reply:
> Allowing IP:port seems bad since it will never work with TLS verification.

It sometimes works, like the DNS-over-HTTPS (e.g. try https://1.1.1.1/). Also the configured dynamic forward proxy may have some requests with IP:port, some with host:port, in the same filter chain.